Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,238 @@
+        Flag indicating whether to store gradients for each function.
+
+    Reference
+    ---------
+    Johnson, R. and Zhang, T., 2013. Accelerating stochastic gradient descent using predictive variance reduction. Advances in neural information processing systems, 26.
+    
+    """
+
+    def __init__(self, functions, sampler=None, update_frequency=None, store_gradients=False):
+        super(SVRGFunction, self).__init__(functions, sampler)
+
+        # update_frequency for SVRG
+        self.update_frequency = update_frequency
+    
+        if self.update_frequency is None:
+            self.update_frequency = 2*self.num_functions
+
+        # compute and store the gradient of each function in the finite sum
+        self.store_gradients = store_gradients
+
+        self._svrg_iter_number = 0
+
+    def gradient(self, x, out=None):
+        """ Selects a random function using the `sampler` and then calls the approximate gradient at :code:`x` or calculates a full gradient depending on the update frequency
+
+        Parameters
+        ----------
+        x : DataContainer
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer
+            the value of the approximate gradient of the sum function at :code:`x` 
+        """
+
+
+    
+
+        if (np.isinf(self.update_frequency) == False and (self._svrg_iter_number % (self.update_frequency)) == 0):
+
+            return self._update_full_gradient_and_return(x, out=out)
+
+        else:
+
+            self.function_num = self.sampler.next()
+
+            if not isinstance(self.function_num, numbers.Number):
+                raise ValueError("Batch gradient is not yet implemented")
+            if self.function_num > self.num_functions:
+                raise IndexError(
+                    'The sampler has outputted an index larger than the number of functions to sample from. Please ensure your sampler samples from {0,1,...,len(functions)-1} only.')
+
+            return self.approximate_gradient(x, self.function_num, out=out)
+
+        
+
+    def approximate_gradient(self, x, function_num, out=None):
+        """ Computes the approximate gradient for each selected function at :code:`x` given a `function_number` in {0,...,len(functions)-1}.
+        
+        Parameters
+        ----------
+        x : DataContainer
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+        Returns
+        --------
+        DataContainer
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+        pass
+    
+
+        self._svrg_iter_number += 1
+
+        self.stoch_grad_at_iterate = self.functions[function_num].gradient(x)
```

Comment:
can we reuse the `stoch_grad_at_iterate`?

```suggestion
        self.stoch_grad_at_iterate = self.functions[function_num].gradient(x, out=self.stoch_grad_at_iterate)
```

Reply:
I tried this but when you pass `out = None` in  `gradient` it doesn't fill the out with what you need and we get an error.